### PR TITLE
Add file type signatures to binary files

### LINKED ISSUE #519: 2.0 file formats need a magic header
http://en.wikipedia.org/wiki/File_format#Magic_number

Would aid Galaxy and other systems

### PR DESCRIPTION
**Pull request by mr-c**
Closes #519

### CODE REVIEW

File: lib/counting.cc [hunk 1]
```diff
@@ -511,11 +511,19 @@ CountingHashFileReader::CountingHashFileReader(
...
+        char signature [14];
         unsigned char version = 0, ht_type = 0, use_bigcount = 0;
 
+        infile.read(signature, 14);
         infile.read((char *) &version, 1);
         infile.read((char *) &ht_type, 1);
-        if (!(version == SAVED_FORMAT_VERSION)) {
+        if (!(std::string(signature) == SAVED_COUNTGRAPH_SIGNATURE)) {
+            std::ostringstream err;
+            err << "Does not start with signature for an Oxli " <<
```
Comment (ctb) [inline lib/counting.cc]: I'm OK with oxli and CountGraph being used internally, but it seems problematic to put them in diagnostic/error output, no?

Comment (mr-c) [inline lib/counting.cc]: Okay. We can change it back during the planned renames.

**Comment by mr-c** [comment]
- [ ] Is it mergeable?
- [ ] Did it pass the tests?
- [ ] If it introduces new functionality in scripts/ is it tested?
  Check for code coverage with `make clean diff-cover`
- [ ] Is it well formatted? Look at `make pep8`, `make diff_pylint_report`,
  `make cppcheck`, and `make doc` output. Use `make format` and manual
  fixing as needed.
- [ ] Did it change the command-line interface? Only additions are allowed
  without a major version increment. Changing file formats also requires a
  major version number increment.
- [ ] Is it documented in the ChangeLog?
  http://en.wikipedia.org/wiki/Changelog#Format
- [ ] Was a spellchecker run on the source code and documentation after
  changes were made?
- [ ] Is the Copyright year up to date?

**Comment by ctb** [comment]
Could you add some discussion about implementation and implications moving forward? How does this break backwards compatibility, etc. etc.?

**Comment by mr-c** [comment]
Discussion in the docs or here?

Old counting table and presence tables will not be loadable with khmer 2.0. Likewise counting table and presence tables created with 2.0 will not load with khmer 1.4 and prior.

**Comment by ctb** [comment]
On Tue, May 26, 2015 at 10:48:22AM -0700, Michael R. Crusoe wrote:
> Discussion in the docs or here?
> 
> Old counting table and presence tables will not be loadable with khmer 2.0. Likewise counting table and presence tables created with 2.0 will not load with khmer 1.4 and prior.

Here is fine.  OK, anything else we should know? :)

**Comment by mr-c** [comment]
To add the header to an old file:

```
echo -n 'OXLICountGraph' > new_file.ct
cat old_counting_table.ct >> new_file.ct

# or 

echo -n 'OXLINodeGraph_' > new_file.pt
cat old_presence_table.ct >> new_file.pt
```

**Comment by luizirber** [comment]
Marginally related to this PR: What is the official procedure to add this signature to libmagic?

**Comment by mr-c** [comment]
@luizirber I just left a comment in the tracking issue https://github.com/dib-lab/khmer/issues/519#issuecomment-105623638

**Comment by mr-c** [comment]
Hmm.. I just noticed that we do include the sub-file type as the second field. I'm going to switch to 'OXLI' as the signature. According to the `file` documentation we just need 32 unique bits, so that should suffice.

**Comment by mr-c** [comment]
- [ ] tagset
- [ ] stoptags
- [ ] subset

**Comment by mr-c** [comment]
@ctb @luizirber ready for review & merge

**Comment by kdmurray91** [comment]
Somewhere in the back of my mind I remember being told to include a non-printable character in the magic string for binary files, as it helps some things (like grep) detect a file is a binary one. I have no idea if it's relevant here, but I know the likes of BAM, gz etc do so.

**Comment by ctb** [comment]
On Tue, May 26, 2015 at 12:44:02PM -0700, Michael R. Crusoe wrote:
> - [ ] tagset
> - [ ] stoptags
> - [ ] subset

also, labels - see #1021.

**Comment by mr-c** [comment]
@kdmurray91 Good point, though our versioning & subtype bytes that follow the magic are not in the printable ASCII range so we are already fine on the point. For example, `file` already correctly identifies a counting table as being of type 'data' prior to the introduction of our rules.

**Comment by mr-c** [comment]
@ctb do you want me to add the signature to your PR or later to this PR for merging after #1021 ?

**Comment by ctb** [comment]
On Wed, May 27, 2015 at 07:27:04AM -0700, Michael R. Crusoe wrote:
> @ctb do you want me to add the signature to your PR or later to this PR for merging after #1021 ?

later, I think.

**Comment by mr-c** [comment]
@ctb  updated for the recently merged labels/tags format. Ready for review & merge

File: doc/whats-new-2.0.rst [hunk 2]
```diff
@@ -0,0 +1,12 @@
+.. vim: set filetype=rst
+
+What's New In khmer 2.0?
+########################
+
+All binary khmer formats (presence tables, counting tables, tag sets, stop tags,
+and partition subsets) has changed. Files are
```
Comment (ctb) [inline doc/whats-new-2.0.rst]: have

File: tests/test_counting_hash.py [hunk 3]
```diff
@@ -323,7 +323,10 @@ def test_save_load():
     hi.save(savepath)
 
     ht = khmer.CountingHash(12, sizes)
-    ht.load(savepath)
+    try:
```
Comment (ctb) [inline tests/test_counting_hash.py]: this try/except should never be triggered unless we're futzing with file formats, right? so... why handle the IOError? Don't we normally expect the tests to pass?

File: tests/test_functions.py [hunk 4]
```diff
@@ -77,12 +77,21 @@ def test_get_primes():
 
 
 def test_extract_countinghash_info():
+    try:
```
Comment (ctb) [inline tests/test_functions.py]: this should be a separate test.

File: tests/test_functions.py [hunk 5]
```diff
@@ -92,11 +101,17 @@ def test_extract_countinghash_info():
...
         try:
             os.remove(fn)
-        except OSError as e:
-            print >>sys.stder, '...failed to remove {fn}'.format(fn)
+        except OSError as err:
+            assert 0, '...failed to remove ' + fn + str(err)
 
 
 def test_extract_hashbits_info():
+    try:
```
Comment (ctb) [inline tests/test_functions.py]: separate test

**Comment by ctb** [comment]
Other than the comments, looks good.  Note that this PR passes the truncation tests I'm adding in #1048 too, so +1 that :)

File: tests/test_counting_hash.py (hunk 3 above)
Comment (mr-c) [inline tests/test_counting_hash.py]: This test errored during development. I make tests fail instead of erroring when it happens to me.

**Comment by mr-c** [comment]
@ctb comments are addressed; ready for review & merge

File: tests/test_counting_hash.py (hunk 3 above)
Comment (ctb) [inline tests/test_counting_hash.py]: OK. I don't regard a trade of four lines for one, with decreased readability and increased complexity, as a fair trade here...

File: .gitattributes [hunk 6]
```diff
@@ -1 +1,2 @@
 khmer/_version.py export-subst
+ChangeLog    merge=merge-changelog
```
Comment (ctb) [inline .gitattributes]: Should this be merged? If so, please add to ChangeLog.

**Comment by ctb** [comment]
Other than the .gitattributes change, LGTM. Please go ahead and merge.

File: .gitattributes (hunk 6 above)
Comment (mr-c) [inline .gitattributes]: Thanks for catching that. Probably shouldn't be merged as not everyone will have git-merge-changelog installed. I've moved it to my personal site-wide configuration file.

Comment (ctb) [inline .gitattributes]: On Sun, Jun 14, 2015 at 12:34:20PM -0700, Michael R. Crusoe wrote:
> > @@ -1 +1,2 @@
> >  khmer/_version.py export-subst
> > +ChangeLog    merge=merge-changelog
> 
> Thanks for catching that. Probably shouldn't be merged as not everyone will have git-merge-changelog installed. I've moved it to my personal site-wide configuration file.

(It doesn't seem to be a problem for merging, note; I ran a merge with it
set, and no problem.)

Comment (mr-c) [inline .gitattributes]: Good to know. Something to investigate for another day.

On Sun, Jun 14, 2015 at 12:36 PM C. Titus Brown <notifications@github.com>
wrote:

> In .gitattributes
> <https://github.com/dib-lab/khmer/pull/1031#discussion_r32383396>:
>
> > @@ -1 +1,2 @@
> >  khmer/_version.py export-subst
> > +ChangeLog    merge=merge-changelog
>
> On Sun, Jun 14, 2015 at 12:34:20PM -0700, Michael R. Crusoe wrote: > @@ -1
> +1,2 @@ > khmer/_version.py export-subst > +ChangeLog merge=merge-changelog
> Thanks for catching that. Probably shouldn't be merged as not everyone will
> have git-merge-changelog installed. I've moved it to my personal site-wide
> configuration file.
> (It doesn't seem to be a problem for merging, note; I ran a merge with it
> set, and no problem.)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/dib-lab/khmer/pull/1031/files#r32383396>.
>
-- 
Michael R. Crusoe: Programmer & Bioinformatician crusoe@ucdavis.edu
<mcrusoe@msu.edu>
The lab for Data Intensive Biology; University of California, Davis
https://impactstory.org/MichaelRCrusoe http://twitter.com/biocrusoe

**Comment by mr-c** [comment]
Jenkins, test this please

**Comment by mr-c** [comment]
@ctb, thanks!

**Pull request merged**